lib/Optimizer/GraphOptimizer/PassManager.cpp

```diff
@@ -0,0 +1,231 @@
+}
+
+/// Global pass counter used to identify each pass.
+static unsigned globalPassCounter = 0;
+
+/// \returns the name of a FunctionPass given its \p passID.
+static llvm::StringRef getNameOfPass(FunctionPassID passID) {
+  switch (passID) {
+#define FUN_PASS(PASS_NAME)                                                    \
+  case FunctionPassID::PASS_NAME:                                              \
+    return #PASS_NAME;
+#include "glow/Optimizer/GraphOptimizer/FunctionPasses.def"
+  }
+}
+
+} // namespace
+
+bool FunctionPassManager::runPrePass(Function *F,
+                                     const CompilationContext &cctx,
+                                     const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Starting Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGBeforeAllPassesOpt ||
+      listContainsString(dumpDAGBeforePassesOpt, P.getName())) {
+    F->dumpDAG(F->getName().str() + "_PrePass_" + P.getName().str() + "_n" +
+               std::to_string(globalPassCounter) + ".dot");
+  }
+  if (verifyBeforeAllPassesOpt ||
+      listContainsString(verifyBeforePassesOpt, P.getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPassManager::runPostPass(Function *F,
+                                      const CompilationContext &cctx,
+                                      const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Finished Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGAfterAllPassesOpt ||
+      listContainsString(dumpDAGAfterPassesOpt, P.getName())) {
+    F->dumpDAG(F->getName().str() + "_PostPass_" + P.getName().str() + "_n" +
+               std::to_string(globalPassCounter) + ".dot");
+  }
+  if (verifyAfterAllPassesOpt ||
+      listContainsString(verifyAfterPassesOpt, P.getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+std::unique_ptr<FunctionPass>
+FunctionPassManager::createFunctionPass(FunctionPassID passID) {
+  switch (passID) {
+#define FUN_PASS(PASS_NAME)                                                    \
+  case (FunctionPassID::PASS_NAME):                                            \
+    return llvm::make_unique<PASS_NAME>();
+#include "glow/Optimizer/GraphOptimizer/FunctionPasses.def"
+  }
+}
+
+bool FunctionPassManager::runPass(const FunctionPassConfig &passConfig,
+                                  Function *F, const CompilationContext &cctx) {
+  const FunctionPassID &passID = passConfig.getFunctionPassID();
+  assert(!(passID == FunctionPassID::DCE &&
+           passConfig.getDCERequiredMode() ==
+               FunctionPassConfig::DCERequiredMode::RequireDCEBefore) &&
+         "Cannot specify DCE requires DCE before it.");
+
+  // Special case for ConstantFold; skip it if specified.
```

Comment:
Is it worth to special case `ConstanFold` here? Why not doing this check inside the pass itself using the provided `cctx`? We have a special-cased DCE already ;-)

Oh, this is probably because passes do not get `cctx` passed to them currently. Is it actually a good idea to not pass a CompilationContext down to passes? What if we have some options that control how a given pass should do its down? How would we inform a pass about it?

Reply:
Sure, probably makes sense to pass the cctx everywhere.